### PR TITLE
Support of i/a typed ternary opcode on P codegen

### PR DESCRIPTION
**Pull request by yanluo7**
Currently ternary opcodes are not widely used, however there are upcoming
optimizer changes in CFG simplification that will generate more integer
and address-typed ternary opcoe (i.e. iternary, lternary, aternary). This
change fixes long ternary (lternary) evaluation on 32 bit platforms where
the result and operands need to be represented by register pairs, in all
various cases handled in evaluator; also, for address-typed ternary
(aternary), it marks the result register(s) as collected reference if
required.

Signed-off-by: Yan Luo <Yan_Luo@ca.ibm.com>

### CODE REVIEW

**Comment by yanluo7** [comment]
@gita-omr Hi Gita, could you review this PR?

**Comment by vijaysun-omr** [comment]
@gita-omr could you please review this ?

File: compiler/p/codegen/ControlFlowEvaluator.cpp [hunk 1]
```diff
@@ -1155,34 +1164,82 @@ TR::Register *OMR::Power::TreeEvaluator::iternaryEvaluator(TR::Node *node, TR::C
...
+            }
          }
       else if (compare_type.isFloatingPoint())
          {
          cmp2Reg = cg->evaluate(firstChild->getSecondChild());
          if (!skip_compare)
             generateTrg1Src2Instruction(cg, TR::InstOpCode::fcmpu, node, ccr, cmp1Reg, cmp2Reg);
-         generateTrg1Src1Instruction(cg, move_opcode, node, resultReg, trueReg);
+
+         if (two_reg)
```
Comment (gita-omr) [inline compiler/p/codegen/ControlFlowEvaluator.cpp]: FloatingPoint will never use two registers (I think) so I would rather remove it. The same at line 1208.

File: compiler/p/codegen/ControlFlowEvaluator.cpp [hunk 2]
```diff
@@ -1155,34 +1164,82 @@ TR::Register *OMR::Power::TreeEvaluator::iternaryEvaluator(TR::Node *node, TR::C
...
+            }
          }
       else
          {
          TR_ASSERT(false, "Unsupported compare type for ternary\n");
          }
 
-      TR::RegisterDependencyConditions *dep = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 5, cg->trMemory());
-      dep->addPostCondition(resultReg, TR::RealRegister::NoReg);
+      TR::RegisterDependencyConditions *dep = two_reg ?
```
Comment (gita-omr) [inline compiler/p/codegen/ControlFlowEvaluator.cpp]: please use 'two_reg ? 8:5' inside new instead of having two news

File: compiler/p/codegen/PPCInstruction.cpp [hunk 3]
```diff
@@ -1571,16 +1571,43 @@ void TR::PPCControlFlowInstruction::assignRegisters(TR_RegisterKinds kindToBeAss
...
+            cg()->traceRAInstruction(cursor = generateTrg1Src1Instruction   (cg(), getOpCode2Value(), currentNode, getTargetRegister(1), getSourceRegister(1), cursor));
+            }
          cg()->traceRAInstruction(cursor = generateConditionalBranchInstruction(cg(), TR::InstOpCode::bne, currentNode, label2, getTargetRegister(0), cursor));
 
-         if (getNumSources() == 4)
+         if (useRegPair())
             {
-            cg()->traceRAInstruction(cursor = generateTrg1Src1ImmInstruction(cg(), TR::InstOpCode::cmpi4, currentNode, getTargetRegister(0), getSourceRegister(3), 0, cursor));
-            cg()->traceRAInstruction(cursor = generateConditionalBranchInstruction(cg(), TR::InstOpCode::bne, currentNode, label2, getTargetRegister(0), cursor));
+            if (getNumSources() == 6)
```
Comment (gita-omr) [inline compiler/p/codegen/PPCInstruction.cpp]: @yanluo7 could you please remind me what is the different number of iternary sources all about (even without pairs).

File: compiler/p/codegen/ControlFlowEvaluator.cpp (hunk 1 above)
Comment (yanluo7) [reply compiler/p/codegen/ControlFlowEvaluator.cpp]: @gita-omr The two_reg flag applies to the 2nd/3rd child nodes where the result could be long value on 32 bit; the isFloatingPoint applies to the 1st compare child which could be a compare node of type float. e.g. 

lternaty
   dcompare
      some regload
      1.23456
   lconst -1
   lconst -2
The result of the ternary is a long (either -1 or -2 based on the double comparison) so on 32bit a reg pair is needed.

File: compiler/p/codegen/ControlFlowEvaluator.cpp (hunk 2 above)
Comment (yanluo7) [reply compiler/p/codegen/ControlFlowEvaluator.cpp]: @gita-omr Fixed.

File: compiler/p/codegen/PPCInstruction.cpp (hunk 3 above)
Comment (yanluo7) [reply compiler/p/codegen/PPCInstruction.cpp]: @gita-omr  This was existing logic, it seems to come from https://github.com/eclipse/omr/blob/6e1c999e6109f80c8e09abe629610fe4b27a66ea/compiler/p/codegen/ControlFlowEvaluator.cpp#L1284-L1285
It says if the result of the first child is a long value and we are on 32 bit, it needs to add the high order of the reg pair to the source register list. In the code above it says if we have the high order register added, in addition to comparing the lower order reg we must also compare the additional reg in order to fully test the condition.

File: compiler/p/codegen/ControlFlowEvaluator.cpp (hunk 1 above)
Comment (gita-omr) [reply compiler/p/codegen/ControlFlowEvaluator.cpp]: got it

File: compiler/p/codegen/PPCInstruction.cpp (hunk 3 above)
Comment (gita-omr) [reply compiler/p/codegen/PPCInstruction.cpp]: Thanks. Sorry for being picky but then it becomes quite confusing what useRegPair() means. Does it mean for the first child or for the last two. We use it for the last two and then decide on the first one based on the number of sources which is really hard to follow inside TR::PPCControlFlowInstruction::assignRegisters()

Maybe we could add two queries: useRegPairForCond() and useRegPairForResult() or somethinglike that...

Comment (yanluo7) [reply compiler/p/codegen/PPCInstruction.cpp]: @gita-omr I agree I think using query makes for better code -- added useRegPairForCond and renamed useRegPairForResult.

**Comment by yanluo7** [comment]
@gita-omr Hi Gita, are you OK with the revision? If so, I can start the merge process. Thank you.

**Comment by andrewcraik** [comment]
@gita-omr ping?

File: compiler/p/codegen/PPCInstruction.cpp [hunk 4]
```diff
@@ -1571,16 +1571,43 @@ void TR::PPCControlFlowInstruction::assignRegisters(TR_RegisterKinds kindToBeAss
          {
          cg()->traceRAInstruction(cursor = generateTrg1Src1ImmInstruction(cg(), TR::InstOpCode::cmpi4, currentNode, getTargetRegister(0), getSourceRegister(0), 0, cursor));
 
-         cg()->traceRAInstruction(cursor = generateTrg1Src1Instruction   (cg(), getOpCode2Value(), currentNode, getTargetRegister(1), getSourceRegister(1), cursor));
+         if (useRegPairForResult())
+            {
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/p/codegen/PPCInstruction.cpp]: Was hoping this can be improved since this change added more variations in child nodes numbering..

[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/p/codegen/PPCInstruction.cpp]: Alright, as long as you double checked we can move it to another item.

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
All checks had passed before @yanluo7 changed the commit message.
So I do not want to re-run full testing.
